### PR TITLE
Mention commit messages in the contribution guidelines.

### PR DESCRIPTION
**Pull request by mkcor**
## Description

@stefanv as per the recent email thread about commit messages.

I wasn't 100% sure about linking to the [reference](https://blogs.vmware.com/opensource/2021/04/14/improve-your-git-commits-in-two-easy-steps/) @alexdesiqueira and you shared, at least in writing (I think it's okay to hear misnomers, but less okay to read them): In the first paragraph, I suspect that "Well-structured and well-written git commits" stands for "Well-structured git commits and well-written git commit messages" and, slightly later, I wonder whether "writing quality git commits" would stand for "writing quality git commit messages."

Actually, I don't think "writing commits" would be incorrect (that would be writing everything the commit consists of, not only its message). Since this talk by Rose Judge deals with commits overall (not just commit messages---which is also the whole point of the talk), it's not 100% clear to me whether "writing quality git commits" is just short for "writing quality git commit messages."

Anyhow, I went reductionist and decided to link to the good ol' post by Chris Beams on *just writing commit messages* (which inevitably impacts how you structure actual commits anyway). But I'm happy to update the PR if the majority prefers the other reference.

I removed the part about merge messages since we now 'squash and merge.'

## Checklist

- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- Gallery example in `./doc/examples` (new features only)
- Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- Unit tests
- Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

## For reviewers

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

### CODE REVIEW

**Comment by stefanv** [comment]
Thank you, Marianne.  I think Chris's post is great.  Unfortunately, it's also a bit of a slog to get through.  I like this summary from Charl Botha's software development guide: https://vxlabs.com/software-development-handbook/#good-commit-messages

Perhaps we should include a similar summary, or link to his?

I am worried that this tiny sentence will get lost in the guidelines.  Perhaps we should also update our [pull request template](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/creating-a-pull-request-template-for-your-repository) to emphasize it?

**Comment by mkcor** [comment]
All very good points, @stefanv. This summary from Charl Botha's software development guide is great and very fitting, indeed. And it starts with a link to Chris's post, so I've happily made the change.

**Review by grlee77 — APPROVED** [review]
Thanks @mkcor

We may also want to mention somewhere in the maintainer guide about editing the messages a bit when squashing the commits. I often delete or condense a number of the less informative ones like "pep8 fix", "apply review comments", etc, while making sure to maintain any Co-authored-by entries.

**Comment by mkcor** [comment]
@grlee77 good point! Done 13e5ae5d825a910d036bc95029b1ee4ae124c384.

File: CONTRIBUTING.txt [hunk 1]
```diff
@@ -60,13 +60,16 @@ Here's the long and short of it:
...
-   * Create a branch for the feature you want to work on. Since the
-     branch name will appear in the merge message, use a sensible name
+   * Create a branch for the feature you want to work on. Use a sensible name,
      such as 'transform-speedups'::
 
       git checkout -b transform-speedups
 
-   * Commit locally as you progress (``git add`` and ``git commit``)
+   * Commit locally as you progress (with ``git add`` and ``git commit``).
+     Please write standard, `good commit messages
```
Comment (stefanv) [inline CONTRIBUTING.txt:69]:
```suggestion
     Please write `good commit messages
```

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -14,6 +14,7 @@
   existing benchmark
 - Unit tests
 - Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
+- Standard commit messages
```
Comment (stefanv) [inline .github/PULL_REQUEST_TEMPLATE.md:17]:
```suggestion
- Descriptive commit messages (see below)
```

**Review by stefanv — APPROVED** [review]
Small tweak, since I don't think most people know what "standard" commit messages are.  Otherwise I think this is good to go!

**Comment by mkcor** [comment]
Thanks, @stefanv! Your latest review allowed me to catch a leftover typo (c74c3bf6a721ba).

File: doc/source/core_developer.md [hunk 3]
```diff
@@ -135,6 +135,12 @@ speak up in case they are not fine with the current state of the pull request.
 Another exceptional situation is one where the first approving review happened
 a long time ago and many changes have taken place in the meantime.
 
+When squashing the commits, please edit the commit messages if necessary, to
+ensure they are standard and they will read nicely in the commit history.
+Typically reorganize them, deleting or condensing the less informative ones
+such as "pep8 fix", "apply review comments", etc. Please make sure to maintain
+all Co-authored-by entries.
+
```
Comment (stefanv) [inline doc/source/core_developer.md:143]: Sorry, @mkcor, I was half asleep yesterday and didn't catch that this paragraph also includes the use of "standard".

```suggestion
When squashing commits Github concatenates all commit messages.  Please edit the resulting message so that it gives a concise, tidy overview of changes.  For example, you may want to grab the description from the PR itself, and delete lines such as "pep8 fix", "apply review comments", etc. Please retain all Co-authored-by entries.
```

File: CONTRIBUTING.txt [hunk 4]
```diff
@@ -60,13 +60,16 @@ Here's the long and short of it:
...
      such as 'transform-speedups'::
 
       git checkout -b transform-speedups
 
-   * Commit locally as you progress (``git add`` and ``git commit``)
+   * Commit locally as you progress (with ``git add`` and ``git commit``).
+     Please write `good commit messages
+     <https://vxlabs.com/software-development-handbook/#good-commit-messages>`_:
+     Capitalize the message, limit it to 50 characters, write it in the
+     imperative mood, etc.
```
Comment (stefanv) [inline CONTRIBUTING.txt:72]:
```suggestion
```

I suggest removing this; it can all be a bit overwhelming, and it is nicely summarized in that blog post (most English-speaking people have to look up imperative mood!).  It's also a bit subtle, because this really only applies to the subject line and not the rest of the commit message.

**Review by stefanv — APPROVED** [review]
I did a half-baked review yesterday.  Apologies for the churn, @mkcor!

File: doc/source/core_developer.md (hunk 3 above)
Comment (mkcor) [reply doc/source/core_developer.md:143]: No worries, it's not too late! :smile: 
I think that if someone is a core developer, they probably know of 'standard' commit messages, so this case is a little different, I would argue. That being said, I like the overall rewording you suggested for this paragraph. About the form:
- I prefer writing 'GitHub' instead of 'Github' (as can be found in the rest of the file);
- I prefer wrapping lines at 79 characters (as done in the rest of the file).

File: CONTRIBUTING.txt (hunk 4 above)
Comment (mkcor) [reply CONTRIBUTING.txt:72]: Ok. In my intention, it was friendlier to expand a little in the body text, but I'm happy to save a couple of lines!

File: doc/source/core_developer.md (hunk 3 above)
Comment (stefanv) [reply doc/source/core_developer.md:143]: :+1: That is the correct capitalization of GitHub, thanks!  And line wrapping is a mess when making suggestions, let me feed it through my editor.  OK, done.

**Comment by stefanv** [comment]
Looks good to me; thanks @mkcor!

File: doc/source/core_developer.md (hunk 3 above)
Comment (mkcor) [reply doc/source/core_developer.md]: Thank you, @stefanv! Syncing up locally, I caught these leftovers: be56fcf726f4ae3a39385489b481aab5b4056a93 (iiuc, it takes two spaces after a period in plain English, but not in Markdown because it's up to the renderer to... render the proper spacing; like LaTeX, right).

Comment (stefanv) [reply doc/source/core_developer.md]: It's an old habit, originating from the time of typewriters, that was taught to me when I was young: back then, the editors/"word processors" didn't space things out automatically, so you added two spaces to create a slightly bigger space between sentences. It isn't necessary anywhere any more; but as they say: old habits die hard.

**Comment by mkcor** [comment]
> Looks good to me; thanks @mkcor!

@stefanv great! Please go ahead and merge :sweat_smile:

**Review by grlee77 — APPROVED** [review]
